src/apps/lwaftr/lwaftr.lua

```diff
@@ -379,9 +380,63 @@ local function icmpv4_incoming(lwstate, pkt)
                            lwstate.aftr_mac_b4_side, lwstate.b4_mac)
 end
 
--- TODO: correctly handle fragmented IPv4 packets
+
+local function cache_ipv4_fragment(lwstate, frag)
+   local cache = lwstate.fragment4_cache
+   local frag_id = rd16(frag.data + lwstate.l2_size + constants.o_ipv4_identification)
+   local src_ip = fstring(frag.data + lwstate.l2_size + constants.o_ipv4_src_addr, 8)
+   local dst_ip = fstring(frag.data + lwstate.l2_size + constants.o_ipv4_dst_addr, 8)
+   local src_dst = src_ip .. dst_ip
+
+   cache[src_dst] = cache[src_dst] or { n = 0 }
+   if not cache[src_dst][frag_id] then
+      cache[src_dst].n = cache[src_dst].n + 1
+      cache[src_dst][frag_id] = {}
+   end
+
+   table.insert(cache[src_dst][frag_id], frag)
+   return cache[src_dst][frag_id]
+end
+
+local function clean_ipv4_fragment_cache(lwstate, frag)
+   local cache = lwstate.fragment4_cache
+   local frag_id = rd16(frag.data + lwstate.l2_size + constants.o_ipv4_identification)
+   local src_ip = fstring(frag.data + lwstate.l2_size + constants.o_ipv4_src_addr, 8)
+   local dst_ip = fstring(frag.data + lwstate.l2_size + constants.o_ipv4_dst_addr, 8)
+   local src_dst = src_ip .. dst_ip
+
+   cache[src_dst][frag_id] = nil
```

Comment:
Compared to the IPv6 fragment cache, I have decided to use *(src, dst)* as first level key, and the fragment identifier (*frag_id*) as the second level. The reason is that there can be multiple sets of fragments between the same source and destination host, with different fragment identifiers, being received in the same time interval. The IPv4 cache is cleaned up with:

```lua
cache[frag_id] = nil
```
which removes *all* the fragments gathered so far with a given fragment identifier, *regardless of the source and the destination*. Which means that, if I understand things correctly, would have the same effect as the packets being dropped/lost, thus increasing the retransmissions done by the source. Let me illustrate with an example on how the existing IPv6 fragment cache would work:

1. A fragment from `A` arrives, destination is `B` and the fragment identifier is `I`:
   * The packet is appended to `cache[I][A..B]`
2. A fragment from `A` arrives, destination is `C` and the fragment identifier is `I`:
   * The packet is appended to `cache[I][A..C]`
3. A fragment from `A` arrives, destination is `B`, and the fragment identifier is `I`. The fragment turns out to be invalid:
   * The cache is cleared with `cache[I] = nil`, which means that:
      * Packets for the `cache[I][A..B]` flow of fragments are dropped (correct).
      * Packets for the `cache[I][A..C]` flow of fragments are dropped (*wrong*).

And that's why I have decided to use the *(src, dst)* pair as first level, and the fragment identifier as second. That way it is possible to clear the fragment cache for a given *(src, dst, id)* set.